hkmc2/shared/src/main/scala/hkmc2/semantics/Resolver.scala

```diff
@@ -46,53 +44,71 @@ object Resolver:
    */
   case class ICtx(
     parent: Opt[ICtx], 
-    iEnv: Map[Type.Sym, Ls[(Type, ICtx.Instance)]],
+    iEnv: Map[Type, Ls[(Type, ICtx.Instance)]],
     tEnv: Map[VarSymbol, Type]
   ):
     
-    def +(typ: Type.Specified, sym: Symbol): ICtx =
-      val newLs = (typ -> ICtx.Instance(sym)) :: iEnv.getOrElse(typ.toSym, Nil)
-      val newEnv = iEnv + (typ.toSym -> newLs)
-      copy(iEnv = newEnv)
+    def +(typ: Type, ins: Symbol): ICtx = typ match
+      case Type.NotImplemented => lastWords("Cannot add instance with a not-yet implemented type.")
+      case Type.Error => this
+      case _ => copy(iEnv = 
+        iEnv + (typ.key -> ((typ -> ICtx.Instance(ins)) :: iEnv.getOrElse(typ.key, Nil)))  
+      )
     
     def withTypeArg(param: VarSymbol, arg: Type): ICtx =
       copy(tEnv = tEnv + (param -> arg))
+    
+    extension (t: Type)
+      private def key: Type = t match
+        case Type.Ref(base, _) => Type.Ref(base, Nil)
+        case _ => t
+    extension (lhs: Type)
+      private def =:= (rhs: Type): Bool = lhs <= rhs && rhs <= lhs
+      private def <= (rhs: Type): Bool = (lhs, rhs) match
```

Comment:
Why choose `=:=` and `<=` instead of `<:<`?

Reply:
IDK. Is it a convention to use `<:<`? I can change it

Reply:
Yes.